File: algoritmos/IV.3 Sequential block/src/python/IV.3-Sequential-block.py
import numpy as np
import sys

def read_matrices_from_file(filename):
    with open(filename, 'r') as f:
        lines = [line.strip() for line in f if line.strip()]
        
        matrix_size = lines.index('') if '' in lines else len(lines) // 2
        n = matrix_size

        A = np.array(
            [[int(num) for num in line.split()] for line in lines[:n]], dtype=np.int64
        )
        
        B = np.array(
            [[int(num) for num in line.split()] for line in lines[n:]], dtype=np.int64
        )
    
    return A, B, n

def sequential_block_multiply(A, B, n, block_size):
    C = np.zeros((n, n), dtype=np.int64)

    for i in range(0, n, block_size):
        for j in range(0, n, block_size):
            for k in range(0, n, block_size):
                for ii in range(i, min(i + block_size, n)):
                    for jj in range(j, min(j + block_size, n)):
                        sum = 0
                        for kk in range(k, min(k + block_size, n)):
                            sum += A[ii, kk] * B[kk, jj]
                        C[ii, jj] += sum
    return C

def print_matrix(M):
    for row in M:
        print(" ".join(map(str, row)))

def main():
    if len(sys.argv) < 2:
        print("Usage: python Sequential-Block.py <matrix_file>")
        sys.exit(1)
    
    filename = sys.argv[1]
    A, B, n = read_matrices_from_file(filename)
    block_size = 2  # You can adjust the block size if needed

    C = sequential_block_multiply(A, B, n, block_size)
    
    print("Result matrix C:")
    print_matrix(C)

if __name__ == "__main__":
    main()

File: algoritmos/IV.3 Sequential block/src/c/IV.3-Sequential-Block.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <inttypes.h>
#include "../../../../utils/matrix_utils.c"  // Include your matrix utility functions

#define BLOCK_SIZE 2  // Block size

// Function to perform sequential block matrix multiplication
void sequential_block_multiply(int64_t **A, int64_t **B, int64_t **C, int n, int block_size) {
    for (int i = 0; i < n; i += block_size) {
        for (int j = 0; j < n; j += block_size) {
            for (int k = 0; k < n; k += block_size) {
                for (int ii = i; ii < i + block_size && ii < n; ii++) {
                    for (int jj = j; jj < j + block_size && jj < n; jj++) {
                        int64_t sum = 0; // Local sum for this block
                        for (int kk = k; kk < k + block_size && kk < n; kk++) {
                            sum += A[ii][kk] * B[kk][jj];
                        }
                        C[ii][jj] += sum; // Accumulate result in C
                    }
                }
            }
        }
    }
}

// Function to print the matrix
void print_matrix(int64_t **M, int n) {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            printf("%" PRId64 " ", M[i][j]);
        }
        printf("\n");
    }
}

// Main function
int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <matrix_file>\n", argv[0]);
        return 1;
    }

    int64_t **A, **B, **C;
    int n;
    const char *filename = argv[1];  // Path to the matrix file

    if (read_matrices_from_file(filename, &A, &B, &n) != 0) {
        fprintf(stderr, "Failed to read matrices from file.\n");
        return -1;
    }

    // Allocate memory for the result matrix C and initialize to zero
    C = (int64_t **)malloc(n * sizeof(int64_t *));
    for (int i = 0; i < n; i++) {
        C[i] = (int64_t *)malloc(n * sizeof(int64_t));
        memset(C[i], 0, n * sizeof(int64_t)); // Initialize to zero
    }

    // Perform sequential block matrix multiplication
    sequential_block_multiply(A, B, C, n, BLOCK_SIZE);

    // Print the result matrix
    printf("Result matrix C:\n");
    print_matrix(C, n);

    // Free allocated memory
    for (int i = 0; i < n; i++) {
        free(A[i]);
        free(B[i]);
        free(C[i]);
    }
    free(A);
    free(B);
    free(C);

    return 0;
}

